src/main/java/org/folio/event/handler/ItemCreateAsyncRecordHandler.java

```diff
@@ -0,0 +1,121 @@
+      log.info("handle:: unsupported event type: {}", eventType);
+      return Future.succeededFuture();
+    }
+
+    if (!validatePayload(payload)) {
+      log.warn("handle:: payload validation failed. payload={}", payload);
+      return Future.succeededFuture();
+    }
+
+    try {
+      var tenantId = extractTenantFromHeaders(kafkaConsumerRecord.headers());
+      var dbClient = new DBClient(getVertx(), tenantId);
+      return processItemCreationEvent(payload, dbClient)
+        .onSuccess(v -> log.info("handle:: item '{}' event processed successfully", eventType))
+        .onFailure(t -> log.error("Failed to process event: {}", kafkaConsumerRecord.value(), t))
+        .map(kafkaConsumerRecord.key());
+    } catch (Exception e) {
+      log.error("Failed to process item event kafka record, kafkaRecord={}", kafkaConsumerRecord, e);
+      return Future.failedFuture(e);
+    }
+  }
+
+  private Future<Void> processItemCreationEvent(JsonObject payload, DBClient dbClient) {
+    var tenantId = payload.getString("tenant");
+    var itemObject = payload.getJsonObject("new");
+    var itemId = itemObject.getString("id");
+
+    return pieceService.getPiecesByItemId(itemId, dbClient)
+      .compose(pieces -> updatePieces(pieces, itemObject, tenantId, dbClient));
+  }
+
+  private Future<Void> updatePieces(List<Piece> pieces, JsonObject itemObject, String tenantId, DBClient client) {
+    if (CollectionUtils.isEmpty(pieces)) {
+      String itemId = itemObject.getString("id");
+      log.info("updatePieces:: no pieces to update found, nothing to update for item={}, tenant={}",
+        itemId, tenantId);
+      return Future.succeededFuture();
+    }
+
+    var holdingId = itemObject.getString("holdingsRecordId");
+    var updateRequiredPieces = filterPiecesToUpdate(pieces, holdingId, tenantId);
+    updatePieceFields(updateRequiredPieces, holdingId, tenantId);
+
+    log.info("updatePieces:: updating '{}' piece(s) out of all '{}' piece(s)",
+      updateRequiredPieces.size(), pieces.size());
+    return pieceService.updatePieces(updateRequiredPieces, client);
+  }
+
+  private List<Piece> filterPiecesToUpdate(List<Piece> pieces, String holdingId, String tenantId) {
+    return pieces.stream()
+      .filter(piece ->
+        ObjectUtils.notEqual(piece.getReceivingTenantId(), tenantId)
+        || ObjectUtils.notEqual(piece.getHoldingId(), holdingId))
+      .filter(piece ->
+        ObjectUtils.notEqual(piece.getReceivingTenantId(), tenantId)
+        || Objects.isNull(piece.getLocationId()))
+      .toList();
+  }
+
+  private void updatePieceFields(List<Piece> pieces, String holdingId, String tenantId) {
+    pieces.forEach(piece -> {
+      piece.setReceivingTenantId(tenantId);
+      if (Objects.isNull(piece.getLocationId())) {
+        piece.setHoldingId(holdingId);
+      }
+    });
+  }
+
+  private boolean validatePayload(JsonObject payload) {
+    var newObject = payload.getJsonObject("new");
```

Comment:
We can avoid creating additional variable here

Reply:
Remove validatoin and moved to handle method since it is small, thanks to notice